### PR TITLE
Support reregistration in worker-runner

### LINKED ISSUE #3012: Add support for reregister endpoint in worker-runner
The implementer of this issue should work closely with the implementer of  #3011.

This will make worker-runner support being handed a token back at `registerWorker` time. The token must be securely stored in a way that makes it inaccessible from tasks. It will pass this token back to worker-manager when it calls `reregisterWorker` and then update the worker with the new creds it gets in return.

This is one part of #2893

### PR DESCRIPTION
**Pull request by djmitche**
Github Bug/Issue: Fixes #3012

### CODE REVIEW

**Comment by djmitche** [comment]
This will need a few rebases, but they should be relatively straightforward.  Let's do a round of review commit-by-commit, starting with "Add locking".

**Comment by petemoore** [comment]
> This will need a few rebases, but they should be relatively straightforward. Let's do a round of review commit-by-commit, starting with "Add locking".

Is there value in splitting this PR into a series of smaller ones?

**Comment by djmitche** [comment]
> Is there value in splitting this PR into a series of smaller ones?

Good idea -- most of these commits are refactors..

**Comment by djmitche** [comment]
Actually only the first commit (a4e18c8) could be broken out.  The reason for the long rebase is that this PR (with the exception of a4e18c8) depends on @helfi92's work which depends on @imbstack's encrypted-column work which we can't land until it gets security review.

File: tools/worker-runner/registration/registration.go [hunk 1]
```diff
@@ -96,22 +98,81 @@ func (reg *RegistrationManager) WorkerStarted() error {
...
 		}
 	})
 
 	return nil
 }
 
+// Request new credentials and send them to the worker, falling back to requesting
+// a worker termination if anything fails.  Called with reg.state locked
+func (reg *RegistrationManager) reregisterWorker() {
+	log.Println("Taskcluster Credentials are expiring in 30s; re-registering")
```
Comment (imbstack) [inline tools/worker-runner/registration/registration.go:119]: I wonder if we should give a bit more headroom than 30s in the case of retries, etc.

File: tools/worker-runner/registration/registration.go [hunk 2]
```diff
@@ -96,22 +98,81 @@ func (reg *RegistrationManager) WorkerStarted() error {
...
+		log.Printf("Could not create worker-manager client: %v", err)
+		reg.terminateWorker()
+		return
+	}
+
+	res, err := wm.ReregisterWorker(&tcworkermanager.ReregisterWorkerRequest{
+		//WorkerPoolID:        reg.state.WorkerPoolID,
+		//ProviderID:          reg.state.ProviderID,
+		//WorkerGroup:         reg.state.WorkerGroup,
+		//WorkerID:            reg.state.WorkerID,
```
Comment (imbstack) [inline tools/worker-runner/registration/registration.go:132]: I haven't looked at the worker-manager patch yet but I'm assuming these shouldn't be commented out?

File: tools/worker-runner/run/state.go [hunk 3]
```diff
@@ -98,3 +102,61 @@ func (state *State) CheckProviderResults() error {
...
+	}
+	err = ioutil.WriteFile(filename, encoded, 0700)
+	if err != nil {
+		return err
+	}
+
+	// This file contains secrets, so ensure that this is really only
+	// accessible to the file owner (and having just created the file, that
+	// should be the current user).
+	err = perms.MakePrivateToOwner(filename)
```
Comment (imbstack) [inline tools/worker-runner/run/state.go:122]: Should we be doing this in a different order? Create empty file -> set permissions -> append content? This leaves it potentially readable for a bit and longer if we crash, right?

File: tools/worker-runner/registration/registration.go (hunk 2 above)
Comment (djmitche) [reply tools/worker-runner/registration/registration.go:132]: Yeah, I'll catch that on rebase, thanks!

File: tools/worker-runner/registration/registration.go (hunk 1 above)
Comment (djmitche) [reply tools/worker-runner/registration/registration.go:119]: Good point -- and we want to leave some leeway for restarts, etc. too.  I'll make a more sophisticated check.

File: tools/worker-runner/run/state.go (hunk 3 above)
Comment (djmitche) [reply tools/worker-runner/run/state.go:122]: Yeah, let's defer that to a followup though -- this is just moving existing code.  #3067.

**Comment by djmitche** [comment]
Oh, another issue: this never resets the timer after reregistering!

**Comment by djmitche** [comment]
Rebased on top of master -- I can rebase onto master again once @helfi92's work lands.

File: services/worker-manager/schemas/v1/reregister-worker-request.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  workerGroup: {$ref: "worker-full.json#/properties/workerGroup"}
+  workerId: {$ref: "worker-full.json#/properties/workerId"}
+  secret:
+    title: Secret
+    type: string
+    pattern: "^[a-zA-Z0-9_-]{44}$"
+    description: |
+      The secret value that was last configured in `registerWorker` (in the case of a newly registerd worker) or
+      `reregisterWorker`.
+      For more information, refer to https://docs.taskcluster.net/docs/reference/core/worker-manager#reregistration.
```
Comment (petemoore) [inline services/worker-manager/schemas/v1/reregister-worker-request.yml:16]: I didn't realise we still maintain `docs.taskcluster.net`.

Comment (djmitche) [reply services/worker-manager/schemas/v1/reregister-worker-request.yml:16]: It's automatically deployed from master.

**Pull request merged**